device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceTwin/DeviceTwin.java

```diff
@@ -403,6 +417,61 @@ public void subscribeDesiredPropertiesTwinPropertyNotification(Map<Property, Pai
         checkSubscription();
     }
 
+    public void getClientProperties(ClientPropertiesCallback clientPropertiesCallback, Object context)
+    {
+        checkSubscription();
+
+        IotHubTransportMessage getTwinRequestMessage = new IotHubTransportMessage(new byte[0], MessageType.DEVICE_TWIN);
+        getTwinRequestMessage.setRequestId(UUID.randomUUID().toString());
+        getTwinRequestMessage.setCorrelationId(getTwinRequestMessage.getRequestId());
+        getTwinRequestMessage.setDeviceOperationType(DeviceOperations.DEVICE_OPERATION_TWIN_GET_REQUEST);
+        getTwinRequestMessage.setClientPropertiesCallback(clientPropertiesCallback);
+        getTwinRequestMessage.setClientPropertiesCallbackContext(context);
+        this.deviceIO.sendEventAsync(getTwinRequestMessage, null, null, this.config.getDeviceId());
+    }
+
+    public synchronized void updateClientProperties(ClientPropertyCollection clientPropertyCollection, Integer version, CorrelatingMessageCallback correlatingMessageCallback, Object correlatingMessageCallbackContext, IotHubEventCallback reportedPropertiesCallback, Object callbackContext) throws IOException
+    {
+        if (clientPropertyCollection == null)
+        {
+            throw new IllegalArgumentException("Reported properties cannot be null");
+        }
+
+        String serializedReportedProperties = config.getPayloadConvention().getPayloadSerializer().serializeToString(clientPropertyCollection);
+
+        if (serializedReportedProperties == null)
+        {
+            return;
+        }
+
+        IotHubTransportMessage clientPropertiesRequest = new IotHubTransportMessage(serializedReportedProperties.getBytes(), MessageType.DEVICE_TWIN);
+        clientPropertiesRequest.setCorrelatingMessageCallback(correlatingMessageCallback);
+        clientPropertiesRequest.setCorrelatingMessageCallbackContext(correlatingMessageCallbackContext);
+        clientPropertiesRequest.setConnectionDeviceId(this.config.getDeviceId());
+
+        // MQTT does not have the concept of correlationId for request/response handling but it does have a requestId
+        // To handle this we are setting the correlationId to the requestId to better handle correlation
+        // whether we use MQTT or AMQP.
+        clientPropertiesRequest.setRequestId(UUID.randomUUID().toString());
```

Comment:
If we aren't saving this UUID anywhere, do we even need to attach it to the outgoing message?

Reply:
This UUID can be accessed via the callbacks I implemented.